src/geom-from-osm.cpp

```diff
@@ -96,27 +96,37 @@ geometry_t create_polygon(osmium::Way const &way)
 }
 
 void create_multilinestring(geometry_t *geom,
-                            osmium::memory::Buffer const &ways)
+                            osmium::memory::Buffer const &ways_buffer,
+                            bool force_multi)
 {
-    auto &mls = geom->set<multilinestring_t>();
-
-    for (auto const &way : ways.select<osmium::Way>()) {
-        linestring_t line;
+    auto ways = ways_buffer.select<osmium::Way>();
+    if (ways.size() == 1 && !force_multi) {
+        auto &line = geom->set<linestring_t>();
+        auto &way = *ways.begin();
         fill_point_list(&line, way.nodes());
-        if (line.size() >= 2U) {
-            mls.add_geometry(std::move(line));
+        if (line.size() < 2U) {
+            geom->reset();
+        }
+    } else {
+        auto &multiline = geom->set<multilinestring_t>();
+        for (auto const &way : ways) {
+            linestring_t line;
+            fill_point_list(&line, way.nodes());
+            if (line.size() >= 2U) {
+                multiline.add_geometry(std::move(line));
+            }
+        }
```

Comment:
There is a corner case here when there are multiple ways in the buffer but only one of them is valid. Then still a multilinestring with a single member is created. It's probably okay, just something we should keep in mind and potentially document.

Reply:
I put it on my todo list.